### DISCUSSION TITLE
empirical analysis discussion between Vibhuti and Jie

### DISCUSSION THREAD

**Discussion by JNing0** [QuickPay-Operational-Performance/Data-and-code#76] [General]
Hi @vibhuti6, I created a discussion for us to talk about data and implementation details.

For starters, could you share with me the latest file with all projects to keep?

**Comment by vibhuti6** [comment]
Thanks Jie, it's the same file as before: https://www.dropbox.com/s/3joumz8260a7xbw/projects_to_keep.csv?dl=0

**Reply by JNing0** [reply]
Actually, to be consistent in our analysis, could you write your data into a csv file and share with me? This way we will be using the same data set to run the regressions. You can just send me the ones with quarterly delays. Thanks!

**Reply by vibhuti6** [reply]
Here it is: https://www.dropbox.com/s/oof0w83uzw8v558/reg_df.csv?dl=0 

The code used to create this dataset is [here](https://github.com/QuickPay-Operational-Performance/Data-and-code/blob/master/codes/reg_data.Rmd), the comments describe what each variable represents.

**Reply by JNing0** [reply]
Wonderful! Thank you!

**Reply by JNing0** [reply]
I have a question about the "delay" column. When a delay entry is NA, what does it mean? That it is the first observation of the project so we cannot compute the delay, correct?

**Comment by JNing0** [comment]
I looked into the nonzero delays and have some concerns about winsorization...

The figure below is the histogram of nonzero delays in the data you shared. I threw out the upper and lower 2.5% of the data to avoid outliers. There is the one-year rounding problem we saw earlier but apart from that, the data looks more or less normal.
<img width="604" alt="image" src="https://user-images.githubusercontent.com/58864118/139592039-4748531d-7109-4d62-9ecf-ece24ca36be7.png">

The figure below is the histogram of winsorized nonzero delays in the data you shared.
<img width="599" alt="image" src="https://user-images.githubusercontent.com/58864118/139592158-54919c71-c791-4a16-b4d4-8bfe84a3446e.png">

Somehow winsorization really condenses the observation at the right end and there is an abnormally amount of observation in the last bin: 236-238 days.

Since winsorization seems to significantly change our dependent variable, I am a bit concerned about using it in the regressions. Shall we just truncate the observed delay at e.g. 5% and use that in the linear regression?

**Reply by vibhuti6** [reply]
Hi Jie,

* Yes, the NA values generally mean its the first observation for the project. But it could also indicate errors in data entry. For example, suppose "2019" is entered as "2091". This would be outside the range of dates allowed in Python/R so when you take the difference of such dates, it gives NA.

* Regarding winsorization, I used the default 5% cutoff for both sides but we can try with 2.5% instead. That may give values similar to what you see with the truncated data. We could also do the analysis using truncated delays, I don't think the results would change much (or at least I hope so).

**Reply by JNing0** [reply]
Yes, let's use the truncated data to run regressions. The empirical distribution of the quarterly delay after winsorization is just too different from the true empirical distribution, with such a huge spike at the right end. It will not affect the logistic regressions but could bias estimates in the linear regression. 

From the regressions I've run, 2.5% cut-off is pretty good. Since we don't have a lot of observation with nonzero quarterly delays, let's use 2.5% to get more statistical power.

**Comment by vibhuti6** [comment]
Hi Jie, my plot for winsorized delays looks somewhat different from yours, not sure why. 

Here's the one with 5% winsorized delay in the data I shared:
![Screen Shot 2021-10-31 at 7 04 44 PM](https://user-images.githubusercontent.com/47764977/139604230-4f7d514b-5ad4-4cba-9b04-4f241a624e1d.png)

Here's the one with 2.5% winsorization instead:
![Screen Shot 2021-10-31 at 7 04 54 PM](https://user-images.githubusercontent.com/47764977/139604253-375241a3-1697-4c23-a71d-b02a7aaac911.png)

Here's the [code](https://github.com/QuickPay-Operational-Performance/Data-and-code/blob/master/codes/wins_delay_plot.Rmd) for reference.

**Reply by JNing0** [reply]
I used 100 breaks in the histogram "hist(windelay.nonzero, breaks=100)". If I did not set "breaks=100" and just do "hist(windelay.nonzero)", I get your first figure (see below).

![image](https://user-images.githubusercontent.com/58864118/139604464-eea58958-425a-4766-9400-487fc069136f.png)

**Reply by vibhuti6** [reply]
Thanks! I think the plots have a similar shape with 2.5% winsorization or truncation, not sure if/why one would be strictly better than the other. We can try both, and then decide which to include in the paper. Don't think results would look very different in either scenario!

**Reply by JNing0** [reply]
Actually the empirical distribution of truncated quarterly delay looks quite different from the winsorized result at 2.5%. Below is the histogram with default breaks, to make it easier to compare with your histogram with 2.5% winsorization. Again, the winsorized data has an abnormally high number of delays in the last bin, altering the actual empirical distribution of the quarterly delay. If you use more breaks to get finer intervals on the histogram for the winsorized delay, I suspect that the majority will be in the last one or two days, as in the case with 5% winsorization.

![image](https://user-images.githubusercontent.com/58864118/139608735-02fe8ffc-b3ae-427f-8987-d42ac29ef763.png)

**Reply by vibhuti6** [reply]
Thanks, Jie. Could you share your code for truncating the data? 

As such, I think winsorization, by definition, would have some peaks at the tails because it is replacing extreme values. The benefit, though, is that it allows us to keep the entire dataset instead of discarding the extreme values altogether.

**Reply by JNing0** [reply]
Winsorization does more than simply replacing "extreme" values. If you compare the histograms of the true data and winsorized data, there are _structural_ changes that cannot be explained by simply replacing the upper and lower 2.5% of the data. The entire distribution is changed. The range of delay is also altered. etc.

So, IMHO, winsorized data should _not_ be used in linear regressions. In fact, could you please do the histogram with 100 breaks for the winsorized data at 2.5% level? I'd like to verify whether it simply replaces the "right end" of the data with the last extreme values, creating a significant point mass within e.g., 2 days of delay, as in the 5% winsorization case.

Here is a sample code for truncation:
```
tmp <- lm.data$Delay[id.nonzero]
# take out all the nonzero delays

LB<- sort(tmp)[round(length(tmp)*0.025)+1]
# lower bound at 2.5%
UB<- sort(tmp)[round(length(tmp)*0.975)+1]
# upper bound at 97.5%

tmp.trun <- tmp[tmp >= LB & tmp <= UB] 
# discard the lower and upper 2.5% of the data to avoid outliers
```

**Reply by vibhuti6** [reply]
Here's the definition from the R package I am using:

 ![Screen Shot 2021-11-01 at 8 20 20 PM](https://user-images.githubusercontent.com/47764977/139759602-c84197a1-39d5-4411-b87b-a924162022f9.png)

I will follow up with the 100 breaks plot. Also, if I understand correctly, you are truncating the data after removing all the zero observations? I winsorized on the entire distribution for our initial analysis.

**Reply by JNing0** [reply]
Ah, that makes sense then. Otherwise it was truly puzzling why it creates such a substantial difference. Yes, I only truncated the nonzero quarterly delays, as we are doing conditional analysis for nonzero delays. If you do winsorization on the nonzero delays, then we should see similar results, correct? Could you do that and compare? 

Given this definition, I still slightly prefer using the truncated data. For example, if we see a huge delay due to wrong entry, then we should discard that data point rather than assume that it equals to the 97.5-percentile of the delay. But it probably won't affect our results, so my preference is "slight".

**Reply by vibhuti6** [reply]
Here's the plot, essentially the same as yours except with peaks at -200 and 500 because of replacement. 
![Screen Shot 2021-11-01 at 11 16 47 PM](https://user-images.githubusercontent.com/47764977/139781454-83ec5393-5bee-4237-b20e-bde379346b0f.png)

**Reply by JNing0** [reply]
Great! Very re-assuring! Thanks!
